### PR TITLE
Fix IDynamicInterfaceCastable with shared generic code

### LINKED ISSUE #72909: Support IDynamicInterfaceCastable with shared generic code
Below snippet will fail with a NullRef while trying to do generic lookup in `IInterfaceCastableImpl`. This is because we didn't provide generic context.

The implementation strategy will be somewhere along the lines of:

* When generating the sealed vtable, add a delta (like we do for fat function pointers) to canonical entries on IDynamicCastable implementations so that we know more is needed at runtime.
* Do something to the file format so that we can express "the index of the interface through which implementation is provided" to cover `[DynIntfCastableImpl] interface IFooImpl : ISomething<object>, ISomething<string> { }` when both ISomething provide default implementations.
* Whenever a IDynamicCastable implementation is looked up at runtime and the found method has the delta bit set, create a thunkpool thunk that puts the "interface through which implementation is provided" in the TLS slot.
* Make a cache for the thunkpool thunks.
* Update the shared generic helper to take the value from the TLS slot.

This is a bunch of work.

It's unclear to me if anyone uses IDynIntfCast with generics. It doesn't work great with generics. E.g. interface variance is broken.

```csharp
using System;
using System.Runtime.InteropServices;

var s = (IInterface<string>)new CastableClass<IInterface<string>, IInterfaceCastableImpl<string>>();
s.GetCookie();

class CastableClass<TInterface, TImpl> : IDynamicInterfaceCastable
{
    RuntimeTypeHandle IDynamicInterfaceCastable.GetInterfaceImplementation(RuntimeTypeHandle interfaceType)
        => typeof(TImpl).TypeHandle;
    bool IDynamicInterfaceCastable.IsInterfaceImplemented(RuntimeTypeHandle interfaceType, bool throwIfNotImplemented)
        => true;
}

interface IInterface<out T>
{
    string GetCookie();
}

[DynamicInterfaceCastableImplementation]
interface IInterfaceCastableImpl<T> : IInterface<T>
{
    string IInterface<T>.GetCookie() => typeof(T).ToString();
}
```

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
Fixes #72909.

Internal team ran into this. Turns out CsWinRT also needs this, but they were silently working around instead of pushing on a fix.

The big problem with this one is that we have an interface call to a default interface method that requires generic context. This means we need some kind of instantiating thunk (since callsite didn't provide generic context because it didn't know it). The normal default interface case uses an instantiating thunk that simply indexes into the interface list of `this`. We know the index of the interface (we don't know the concrete type because `T`s could be involved), but we can easily compute it at runtime from `this`.

The problem with `IDynamicInterfaceCastable` is that `this` is useless (the class doesn't know anything about the interface). So we need to get the generic context from somewhere else. In this PR, I'm using the thunkpool as "somewhere else". When we finish interface lookup and find out `IDynamicInterfaceCastable` provided a shared method, we create a thunkpool thunk that stashes away the context. We then call the "default interface method instantiating thunk" and instead of indexing into interface list of `this`, we index into interface list of whatever was stashed away. So there are two thunks before we reach the point of executing the method body.

Submitting as a draft. This needs one more thing - a cache of all the thunks we already generated. I just wanted to run the general approach by everyone before making it even more complicated. This is already too complicated for my taste so maybe someone has a better idea.

Cc @dotnet/ilc-contrib @dotnet/interop-contrib

### CODE REVIEW

**Comment by SingleAccretion** [comment]
I see there is a certain amount of code added to support downstream (WASM). The dynamic instantiation thunk approach won't work on WASM - the thunk pool doesn't work on WASM.

Does implementing this fundamentally require the ability to generate new code at runtime, like with `Delegate.GetFunctionPointerForDelegate(closedDelegate)`?

**Comment by MichalStrehovsky** [comment]
> I see there is a certain amount of code added to support downstream (WASM). The dynamic instantiation thunk approach won't work on WASM - the thunk pool doesn't work on WASM.

Yes, I mindlessly copied the fat function pointer indication bit since dispatch maps never have fat function pointers (if they had, this problem wouldn't exist).

> Does implementing this fundamentally require the ability to generate new code at runtime, like with `Delegate.GetFunctionPointerForDelegate(closedDelegate)`?

The problem is that we need to inject an extra parameter into method calls. I don't see any way around that. CoreCLR JITs new instantiation thunks on demand. We can't.

File: src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/IDynamicInterfaceCastableSupport.cs [hunk 1]
```diff
@@ -28,11 +35,32 @@ internal static IntPtr IDynamicCastableGetInterfaceImplementation(IDynamicInterf
...
-            IntPtr result = RuntimeImports.RhResolveDispatchOnType(implType, interfaceType, slot);
+
+            MethodTable* genericContext = null;
+            IntPtr result = RuntimeImports.RhResolveDynamicInterfaceCastableDispatchOnType(implType, interfaceType, slot, &genericContext);
             if (result == IntPtr.Zero)
             {
                 IDynamicCastableGetInterfaceImplementationFailure(instance, interfaceType, implType);
             }
+
+            if (genericContext != null)
```
Comment (jkotas) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/IDynamicInterfaceCastableSupport.cs:46]: This looks same as the delegate thunk pool. Should we just have one shared thunk pool?

**Review by jkotas — COMMENTED** [review]
> maybe someone has a better idea.

I cannot think about anything better.

File: src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/IDynamicInterfaceCastableSupport.cs (hunk 1 above)
Comment (MichalStrehovsky) [reply src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/IDynamicInterfaceCastableSupport.cs:46]: It feels more risky to touch the delegate thunk pool. We might need to service this to .NET 9. With that in mind would you still prefer we use the same pool?

Comment (jkotas) [reply src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/IDynamicInterfaceCastableSupport.cs:46]: I do not have strong opinion. We can clean it up later. There is quite a bit leftovers from when the thunk pool was in the shared MRT.

**Comment by jkotas** [comment]
Is this going to introduce measurable binary size regression for typical apps on Apple platforms? I would expect that the thunk pool and the associated templates were not linked into basic apps on Apple platforms before this change, and it is going to be linked in most of the time after this change.

(I tried submitting https://github.com/MichalStrehovsky/rt-sz job to get some binary size numbers. It does not seem to work.)

**Review by jkotas — APPROVED** [review]

**Comment by MichalStrehovsky** [comment]
> (I tried submitting https://github.com/MichalStrehovsky/rt-sz job to get some binary size numbers. It does not seem to work.)

It was broken due to .NET 9 -> .NET 10 renaming. Tweaked it to represent current point it time but it will get broken in the coming weeks again.

> Is this going to introduce measurable binary size regression for typical apps on Apple platforms? I would expect that the thunk pool and the associated templates were not linked into basic apps on Apple platforms before this change, and it is going to be linked in most of the time after this change.

Would we expect that to be disproportionate compared to non-Apple platforms (they have the same problem)? Looks like this is single digit kB outside Apple. rt-sz doesn't measure Apple platforms because that support is a PITA to maintain and usually corresponds to Linux anyway. Cc @ivanpovazan @matouskozak FYI if we see a regression in automation after this merges.

**Comment by jkotas** [comment]
> Would we expect that to be disproportionate compared to non-Apple platforms (they have the same problem)?

Yes, because of this: https://github.com/dotnet/runtime/blob/main/src/coreclr/nativeaot/Runtime/Full/CMakeLists.txt#L9-L15

The static binary cost of `FEATURE_RX_THUNKS` is in single digit kB. The static binary cost of `!FEATURE_RX_THUNKS` that is used on Apple platforms is much more. It may or may not show up in the apps that we care about. It depends on whether something else paid the static cost in the app.

**Comment by jkotas** [comment]
> here is a regression for nativeAOT on iOS of around ~600 kB https://github.com/dotnet/perf-autofiling-issues/issues/42390.

Did you mean to say ~60 kB?

**Comment by matouskozak** [comment]
> > here is a regression for nativeAOT on iOS of around ~600 kB [dotnet/perf-autofiling-issues#42390](https://github.com/dotnet/perf-autofiling-issues/issues/42390).
> 
> Did you mean to say ~60 kB?

I did, yes *only* 60kB, sorry. 600 kB would be too much :)
image

**Comment by agocke** [comment]
Seems like the relevant number is 2% of app size, which does not feel particularly impactful.

**Pull request merged**